File: alds1/9c_priority_queue.c
#define dbg(fmt,...) fprintf(stderr,fmt,__VA_ARGS__)
#define dpri(x) dbg(#x ": %d\n", x)
#define dprs(x) dbg(#x ": %s\n", x)
#include <stdio.h>
#include <string.h>
typedef long long ll;
const int MYINF = 1e9+7;
typedef int bool;
const bool true = 1;
const bool false = 0;

#define MAX_H 32000000
#define parent(i) ((i)/2)
#define left(i) ((i)*2)
#define right(i) ((i)*2+1)

int H;
int heap[MAX_H];
#define len heap[0]

void swap(int i, int j) {
	int t;
	t = heap[i];
	heap[i] = heap[j];
	heap[j] = t;
}

void shiftup(int i) {
	int p = parent(i);
	if (i == 1 || heap[p] > heap[i]) return;
	swap(p, i);
	shiftup(p);
}

void insert(int x) {
	heap[++len] = x;
	shiftup(len);
}

void shiftdown(int i) {
	int l = left(i), r = right(i), c = -1000;
	if (len < i) return;
	if (len < l) return;
	if (r <= len) {
		if (heap[l] > heap[i] && heap[r] > heap[i]) {
			if (heap[l] > heap[r]) c = l;
			else c = r;
		} else if (heap[l] > heap[i]) {
			c = l;
		} else if (heap[r] > heap[i]) {
			c = r;
		} else {
			return;
		}
	} else if (l <= len) {
		if (heap[l] > heap[i]) c = l;
		else return;
	}
	swap(c, i);
	shiftdown(c);
}

int extract(void) {
	int res;
	res = heap[1];
	heap[1] = heap[len--];
	shiftdown(1);
	return res;
}

int main(int argc, char **argv)
{
	int x;
	char cmd[256];

	while (1) {
		scanf("%s", cmd);
		if (!strcmp(cmd, "end")) break;
		if (!strcmp(cmd, "insert")) {
			scanf("%d", &x);
			insert(x);
		} else if (!strcmp(cmd, "extract")) {
			printf("%d\n", extract());
		}
	}

	return 0;
}

File: alds1/9c_priority_queue.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"unicode"
)

type Heap struct {
	a []int
}

func NewHeap() *Heap {
	return &Heap{make([]int, 1)}
}

func parent(i int) int {
	return i / 2
}

func left(i int) int {
	return i * 2
}

func right(i int) int {
	return i*2 + 1
}

func (h *Heap) Insert(k int) {
	var shiftUp func(int)
	shiftUp = func(i int) {
		p := parent(i)
		if p <= 0 || h.a[p] > h.a[i] {
			return
		}
		h.a[p], h.a[i] = h.a[i], h.a[p]
		shiftUp(p)
	}
	h.a = append(h.a, k)
	shiftUp(len(h.a) - 1)
}

func (h *Heap) ExtractMax() int {
	var maxHeapify func(int)
	maxHeapify = func(i int) {
		l := left(i)
		r := right(i)
		n := len(h.a) - 1
		largest := -1

		if l <= n && h.a[l] > h.a[i] {
			largest = l
		} else {
			largest = i
		}
		if r <= n && h.a[r] > h.a[largest] {
			largest = r
		}

		if largest != i {
			h.a[i], h.a[largest] = h.a[largest], h.a[i]
			maxHeapify(largest)
		}
	}
	res := h.a[1]
	n := len(h.a) - 1
	h.a[1] = h.a[n]
	h.a = h.a[:n]
	maxHeapify(1)
	return res
}

func main() {
	defer stdout.Flush()
	h := NewHeap()
	for {
		cmd := readString()
		switch cmd {
		case "insert":
			k := readInt()
			h.Insert(k)
		case "extract":
			k := h.ExtractMax()
			println(k)
		case "end":
			return
		default:
			panic("unknown command")
		}
	}
}

var (
	stdin  *bufio.Reader
	stdout *bufio.Writer
)

func init() {
	stdin = bufio.NewReader(os.Stdin)
	stdout = bufio.NewWriter(os.Stdout)
}

func scanf(f string, args ...interface{}) (int, error) {
	n, err := fmt.Fscanf(stdin, f, args...)
	for {
		r, _, err := stdin.ReadRune()
		if err != nil || !unicode.IsSpace(r) {
			stdin.UnreadRune()
			break
		}
	}
	return n, err
}

func readInt() int {
	var n int
	scanf("%d", &n)
	return n
}

func readIntSlice(n int) []int {
	a := make([]int, n)
	for i := 0; i < n; i++ {
		scanf("%d", &a[i])
	}
	return a
}

func readLengthAndSlice() []int {
	n := readInt()
	return readIntSlice(n)
}

func readString() string {
	var s string
	scanf("%s", &s)
	return s
}

func printf(f string, args ...interface{}) (int, error) {
	return fmt.Fprintf(stdout, f, args...)
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdout, args...)
}
